cdc/redo/common/redo.go

```diff
@@ -52,7 +57,54 @@ const (
 
 // LogMeta is used for store meta info.
 type LogMeta struct {
-	CheckPointTs   uint64           `msg:"checkPointTs"`
-	ResolvedTs     uint64           `msg:"resolvedTs"`
-	ResolvedTsList map[int64]uint64 `msg:"-"`
+	CheckPointTs uint64 `msg:"checkPointTs"`
+
+	ResolvedTsList map[int64]uint64 `msg:"resolvedTsList"`
+}
+
+// ResolvedTs returns the minimal resolved timestamp of all tables.
+// If no resolved timestamp is record, return checkpoint instead.
+func (z LogMeta) ResolvedTs() model.Ts {
+	var minResolvedTs uint64 = math.MaxUint64
+	for _, ts := range z.ResolvedTsList {
+		if ts < minResolvedTs {
+			minResolvedTs = ts
+		}
+	}
+	if minResolvedTs == math.MaxUint64 {
+		return z.CheckPointTs
+	}
+	return minResolvedTs
+}
+
+// MarshalMsg implements msgp.Marshaler
+func (z LogMeta) MarshalMsg(b []byte) ([]byte, error) {
```

Comment:
There may exist a compatibility issue here, because previously we use msgpack to encode metadata, but now the metadata is encoded by this customized `MarshalMsg` function. So when users need to apply redolog, could the redo reader decode the old msgpack metadata correctly?

Reply:
Old msgpack data can't be decoded correctly.

The reason to avoid msgpack is seems its map can only support string keys:
> Maps must have string keys. This is intentional (as it preserves JSON interop.) Although non-string map keys are not forbidden by the MessagePack standard, many serializers impose this restriction. (It also means any well-formed struct can be de-serialized into a map[string]interface{}.) The only exception to this rule is that the deserializers will allow you to read map keys encoded as bin types, due to the fact that some legacy encodings permitted this. (However, those values will still be cast to Go strings, and they will be converted to str types when re-encoded. It is the responsibility of the user to ensure that map keys are UTF-8 safe in this case.) The same rules hold true for JSON translation.